### ISSUE TITLE
Multi-Gazebo Kobukis not updating orientations

### ISSUE THREAD

**Issue by jihoonl** [yujinrobot/kobuki_desktop#38]
KGP - Kobuki Gazebo Plugin
- Start two gazebo kobukis (segbot + kobuki is ok)
- Send commands to cmd_vel of one kobuki
- KGP catches these and moves around properly in gazebo world
- **KGP does not post rotations to the odom topic (translations are there)**
- **KGP does not publish rotation information on base_footprint-odom tf.**

**Comment by jihoonl** [comment]
This issue is related.

http://answers.gazebosim.org/question/4878/multiple-robots-with-ros-plugins-sensor-plugin-vs/

**Comment by slivingston** [comment]
#31 that I opened in May 2014 is also related.

**Comment by stonier** [comment]
- Kobuki is manually calculating its differential drive output
- There is a [gazebo diff drive plugin](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/hydro-devel/gazebo_plugins/src/gazebo_ros_diff_drive.cpp) - should we be using that?

**Comment by bit-pirate** [comment]
> There is a gazebo diff drive plugin - should we be using that?

If you do that, then you need to split things up. I.e. develop a second plugin which handles Kobuki's sensor I/O.

**Comment by stonier** [comment]
Historical point : we didn't use that diff drive plugin because it wasn't there when the kobuki gazebo plugin was made.

In conclusion: doesn't sound like work worth chasing that gives us any extra benefits.

**Comment by slivingston** [comment]
@jihoonl , I think the correction is to use `tf::Quaternion::setRPY()` instead of `setEuler()`. Try applying the patch below. Note there are many different conventions in use for "Euler angles", which makes the `setEuler()` function somewhat ambiguous. Quick links to relevant tf API documentation are:
- [setRPY](http://docs.ros.org/api/tf/html/c++/classtf_1_1Quaternion.html#ab34dfef6a01c1ece07bea92d485b0576)
- [setEuler](http://docs.ros.org/api/tf/html/c++/classtf_1_1Quaternion.html#aa64600b31d011dbb00dcb151d2d3d241)

For comparison, the differential drive plugin in the gazebo_ros_pkgs repository [uses setRPY](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/d32c4a3b5154df99d7b4b467658037209923a28d/gazebo_plugins/src/gazebo_ros_diff_drive.cpp#L314). The preceding link is for the tip of [indigo-devel branch](https://github.com/ros-simulation/gazebo_ros_pkgs/tree/indigo-devel) of gazebo_ros_pkgs at time of writing.

I have tested the below patch for the case of spawning two Kobuki models without using the rocon infrastructure.

``` diff
diff --git a/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_updates.cpp b/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_updates.cpp
index 81340fc..17ac762 100644
--- a/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_updates.cpp
+++ b/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_updates.cpp
@@ -105,7 +105,7 @@ void GazeboRosKobuki::updateOdometry(common::Time& step_time)
   odom_.pose.pose.position.z = 0;

   tf::Quaternion qt;
-  qt.setEuler(0,0,odom_pose_[2]);
+  qt.setRPY(0,0,odom_pose_[2]);
   odom_.pose.pose.orientation.x = qt.getX();
   odom_.pose.pose.orientation.y = qt.getY();
   odom_.pose.pose.orientation.z = qt.getZ();
```

**Comment by jihoonl** [comment]
@slivingston Thanks. I will test and get back to the issue.

**Comment by slivingston** [comment]
Update: The above patch is not sufficient, and indeed, we would not expect it to be because there is apparently correct behavior in the case of a single Kobuki. I have nearly identified the bug and will report soon.

**Comment by jihoonl** [comment]
Yeah I have just tested the patch is not sufficient to fix this issue.

To test this in easy way with concert,

[Example Concert : Office sim Solution](https://github.com/robotics-in-concert/rocon_demos/tree/gazebo_concert/office_sim_solution)
1. gazebo_concert with two turtlebot.
2. add [concert_service_indoor_2d_map_prep/indoor_2d_map_prep](https://github.com/robotics-in-concert/concert_services/tree/indigo/concert_service_indoor_2d_map_prep) service.[The up-to-date source as of now!]
3. run Concert Make a Map through rocon_remocon

**Comment by slivingston** [comment]
The bug is due to the different Kobuki plugins using the same ImuSensor. I am creating a patch now.

**Comment by jihoonl** [comment]
Awesome Thanks! :)

**Comment by slivingston** [comment]
``` diff
diff --git a/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_loads.cpp b/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_loads.cpp
index 3e0ef50..5bf4ff1 100644
--- a/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_loads.cpp
+++ b/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_loads.cpp
@@ -295,7 +295,7 @@ bool GazeboRosKobuki::prepareIMU()
     return false;
   }
   imu_ = boost::dynamic_pointer_cast<sensors::ImuSensor>(
-            sensors::SensorManager::Instance()->GetSensor(imu_name));
+     sensors::get_sensor(world_->GetName()+"::"+node_name_+"::base_footprint::"+imu_name));
   if (!imu_)
   {
     ROS_ERROR_STREAM("Couldn't find the IMU in the model! [" << node_name_ <<"]");
```

The basic problem is that the sensor name, as obtained on [line 289 of gazebo_ros_kobuki_loads.cpp](https://github.com/yujinrobot/kobuki_desktop/blob/06d8fa710928ba5623939ce42f760838cd0f4667/kobuki_gazebo_plugins/src/gazebo_ros_kobuki_loads.cpp#L289) is not scoped. Thus it is the same when multiple models are spawned using the same URDF file, namely kobuki_gazebo.urdf.xacro, which declares the imu name [at line 186](https://github.com/yujinrobot/kobuki/blob/06e60c2bba4c73699f69ffc61250f1ad152a931a/kobuki_description/urdf/kobuki_gazebo.urdf.xacro#L186). Hence, the sensor obtained by `sensors::SensorManager::Instance()->GetSensor()` is subject to a race condition. I propose avoiding this by using the above patch, which gets the sensor using a fully scoped name that includes the name of the model associated with the Kobuki Gazebo plugin.

The relevant Gazebo API documentation is [gazebo::sensors::get_sensor()](http://osrf-distributions.s3.amazonaws.com/gazebo/api/2.2.1/group__gazebo__sensors.html#gac47a5236e475a51bdad880486fdb9687).

**Comment by jihoonl** [comment]
@slivingston Could you make a pull request?

**Comment by slivingston** [comment]
I just opened https://github.com/yujinrobot/kobuki_desktop/pull/39

**Comment by jihoonl** [comment]
This has been resolved witih #39
